// @danielx/civet@0.11.7/dist/esm.mjs
// source/esm.civet
import { readFile } from "fs/promises";
import { dirname } from "path";
import { pathToFileURL, fileURLToPath } from "url";
import { compile } from "./main.mjs";
import { findConfig, loadConfig } from "./config.mjs";
var baseURL = pathToFileURL(process.cwd() + "/").href;
var extensionsRegex = /\.civet(?:$|[?#])/;
var globalOptions = {};
function initialize(options = {}) {
  return globalOptions = options;
}
function resolve(specifier, context, next) {
  const { parentURL = baseURL } = context;
  if (extensionsRegex.test(specifier)) {
    return {
      shortCircuit: true,
      format: "civet",
      url: new URL(specifier, parentURL).href
    };
  }
  return next(specifier, context);
}
async function load(url, context, next) {
  if (context.format === "civet") {
    const path = fileURLToPath(url);
    const source = await readFile(path);
    let loadedConfig;
    let { config } = globalOptions;
    if (config === void 0) {
      config = await findConfig(dirname(path));
    }
    if (config) {
      loadedConfig = await loadConfig(config);
    }
    const options = {
      filename: path,
      inlineMap: true,
      js: true
    };
    if (globalOptions.parseOptions) {
      options.parseOptions = globalOptions.parseOptions;
    }
    if (loadedConfig?.parseOptions) {
      options.parseOptions = { ...options.parseOptions, ...loadedConfig.parseOptions };
    }
    let code;
    try {
      code = await compile(source, options);
    } catch (e) {
      console.error(`Civet failed to compile ${url}:`, e);
      throw e;
    }
    return {
      shortCircuit: true,
      format: "module",
      source: code
    };
  }
  return next(url, context);
}
export {
  initialize,
  load,
  resolve
};
