// @danielx/civet@0.11.7/dist/unplugin/unplugin.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// source/unplugin/unplugin.civet
var unplugin_exports = {};
__export(unplugin_exports, {
  default: () => unplugin_default,
  rawPlugin: () => rawPlugin,
  slash: () => slash
});
module.exports = __toCommonJS(unplugin_exports);
var import_unplugin = require("unplugin");
var import_civet = __toESM(require("@danielx/civet"));
var import_config = require("@danielx/civet/config");
var import_ts_diagnostic = require("@danielx/civet/ts-diagnostic");
var fs = __toESM(require("fs"));
var import_path = __toESM(require("path"));

// node_modules/.pnpm/@typescript+vfs@1.6.4_typescript@5.8.3/node_modules/@typescript/vfs/dist/vfs.esm.js
function _extends() {
  return _extends = Object.assign ? Object.assign.bind() : function(n) {
    for (var e = 1; e < arguments.length; e++) {
      var t = arguments[e];
      for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]);
    }
    return n;
  }, _extends.apply(null, arguments);
}
var hasLocalStorage = false;
try {
  hasLocalStorage = typeof localStorage !== "undefined";
} catch (error) {
}
var hasProcess = typeof process !== "undefined";
var shouldDebug = hasLocalStorage && typeof localStorage.getItem === "function" && /* @__PURE__ */ localStorage.getItem("DEBUG") || hasProcess && process.env.DEBUG;
var debugLog = shouldDebug ? console.log : function(_message) {
  return "";
};
function notImplemented(methodName) {
  throw new Error("Method '" + methodName + "' is not implemented.");
}
function audit(name, fn) {
  return function() {
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    var res = fn.apply(void 0, args);
    var smallres = typeof res === "string" ? res.slice(0, 80) + "..." : res;
    debugLog.apply(void 0, ["> " + name].concat(args));
    debugLog("< " + smallres);
    return res;
  };
}
var defaultCompilerOptions = function defaultCompilerOptions2(ts) {
  return _extends({}, ts.getDefaultCompilerOptions(), {
    jsx: ts.JsxEmit.React,
    strict: true,
    esModuleInterop: true,
    module: ts.ModuleKind.ESNext,
    suppressOutputPathCheck: true,
    skipLibCheck: true,
    skipDefaultLibCheck: true
  }, ts.versionMajorMinor && Number(ts.versionMajorMinor.split(".")[0]) >= 6 ? {
    ignoreDeprecations: "6.0"
  } : {
    moduleResolution: ts.ModuleResolutionKind.NodeJs
  });
};
function createFSBackedSystem(files, _projectRoot, ts, tsLibDirectory) {
  var root = _projectRoot + "/vfs";
  var path2 = requirePath();
  var nodeSys = ts.sys;
  var tsLib = tsLibDirectory != null ? tsLibDirectory : path2.dirname(require.resolve("typescript"));
  return {
    // @ts-ignore
    name: "fs-vfs",
    root,
    args: [],
    createDirectory: function createDirectory() {
      return notImplemented("createDirectory");
    },
    // TODO: could make a real file tree
    directoryExists: audit("directoryExists", function(directory) {
      return Array.from(files.keys()).some(function(path3) {
        return path3.startsWith(directory);
      }) || nodeSys.directoryExists(directory);
    }),
    exit: nodeSys.exit,
    fileExists: audit("fileExists", function(fileName) {
      if (files.has(fileName)) return true;
      if (fileName.includes("tsconfig.json") || fileName.includes("tsconfig.json")) return false;
      if (fileName.startsWith("/lib")) {
        var tsLibName = tsLib + "/" + fileName.replace("/", "");
        return nodeSys.fileExists(tsLibName);
      }
      return nodeSys.fileExists(fileName);
    }),
    getCurrentDirectory: function getCurrentDirectory() {
      return root;
    },
    getDirectories: nodeSys.getDirectories,
    getExecutingFilePath: function getExecutingFilePath() {
      return notImplemented("getExecutingFilePath");
    },
    readDirectory: audit("readDirectory", function() {
      if ((arguments.length <= 0 ? void 0 : arguments[0]) === "/") {
        return Array.from(files.keys());
      } else {
        return nodeSys.readDirectory.apply(nodeSys, arguments);
      }
    }),
    readFile: audit("readFile", function(fileName) {
      if (files.has(fileName)) return files.get(fileName);
      if (fileName.startsWith("/lib")) {
        var tsLibName = tsLib + "/" + fileName.replace("/", "");
        var result = nodeSys.readFile(tsLibName);
        if (!result) {
          var libs = nodeSys.readDirectory(tsLib);
          throw new Error("TSVFS: A request was made for " + tsLibName + " but there wasn't a file found in the file map. You likely have a mismatch in the compiler options for the CDN download vs the compiler program. Existing Libs: " + libs + ".");
        }
        return result;
      }
      return nodeSys.readFile(fileName);
    }),
    resolvePath: function resolvePath(path3) {
      if (files.has(path3)) return path3;
      return nodeSys.resolvePath(path3);
    },
    newLine: "\n",
    useCaseSensitiveFileNames: true,
    write: function write() {
      return notImplemented("write");
    },
    writeFile: function writeFile(fileName, contents) {
      files.set(fileName, contents);
    },
    deleteFile: function deleteFile(fileName) {
      files["delete"](fileName);
    },
    realpath: nodeSys.realpath
  };
}
function createVirtualCompilerHost(sys, compilerOptions, ts) {
  var sourceFiles = /* @__PURE__ */ new Map();
  var save = function save2(sourceFile) {
    sourceFiles.set(sourceFile.fileName, sourceFile);
    return sourceFile;
  };
  var vHost = {
    compilerHost: _extends({}, sys, {
      getCanonicalFileName: function getCanonicalFileName(fileName) {
        return fileName;
      },
      getDefaultLibFileName: function getDefaultLibFileName() {
        return "/" + ts.getDefaultLibFileName(compilerOptions);
      },
      // '/lib.d.ts',
      // getDefaultLibLocation: () => '/',
      getNewLine: function getNewLine() {
        return sys.newLine;
      },
      getSourceFile: function getSourceFile(fileName, languageVersionOrOptions) {
        var _ref;
        return sourceFiles.get(fileName) || save(ts.createSourceFile(fileName, sys.readFile(fileName), (_ref = languageVersionOrOptions != null ? languageVersionOrOptions : compilerOptions.target) != null ? _ref : defaultCompilerOptions(ts).target, false));
      },
      useCaseSensitiveFileNames: function useCaseSensitiveFileNames() {
        return sys.useCaseSensitiveFileNames;
      }
    }),
    updateFile: function updateFile(sourceFile) {
      var alreadyExists = sourceFiles.has(sourceFile.fileName);
      sys.writeFile(sourceFile.fileName, sourceFile.text);
      sourceFiles.set(sourceFile.fileName, sourceFile);
      return alreadyExists;
    },
    deleteFile: function deleteFile(sourceFile) {
      var alreadyExists = sourceFiles.has(sourceFile.fileName);
      sourceFiles["delete"](sourceFile.fileName);
      sys.deleteFile(sourceFile.fileName);
      return alreadyExists;
    }
  };
  return vHost;
}
var requirePath = function requirePath2() {
  return require(String.fromCharCode(112, 97, 116, 104));
};

// source/unplugin/unplugin.civet
var import_os = __toESM(require("os"));
var DEFAULT_EXTENSIONS = [".mjs", ".js", ".mts", ".ts", ".jsx", ".tsx", ".json"];
var DiagnosticCategory = {};
DiagnosticCategory[DiagnosticCategory["Warning"] = 0] = "Warning";
DiagnosticCategory[DiagnosticCategory["Error"] = 1] = "Error";
DiagnosticCategory[DiagnosticCategory["Suggestion"] = 2] = "Suggestion";
DiagnosticCategory[DiagnosticCategory["Message"] = 3] = "Message";
var queryPostfixRE = /\?.*$/s;
var escapedHashRE = /\0#/g;
var isWindows = import_os.default.platform() === "win32";
var windowsSlashRE = /\\/g;
var civetSuffix = ".civet";
var workerRE = /(?:\?|&)(?:worker|sharedworker)(?:&|$|#)/;
function extractCivetFilename(id, outputExtension) {
  let postfix = "";
  id = id.replace(escapedHashRE, "#");
  let filename = id.replace(queryPostfixRE, (match) => {
    postfix = match;
    return "";
  });
  if (filename.endsWith(outputExtension)) {
    filename = filename.slice(0, -outputExtension.length);
  }
  const hashIndex = filename.lastIndexOf("#");
  if (hashIndex >= 0 && !tryStatSync(filename)) {
    postfix = filename.slice(hashIndex) + postfix;
    filename = filename.slice(0, hashIndex);
    if (filename.endsWith(outputExtension)) {
      filename = filename.slice(0, -outputExtension.length);
    }
  }
  return { filename, postfix };
}
function tryStatSync(file) {
  try {
    return fs.statSync(file, { throwIfNoEntry: false });
  } catch (e) {
    return;
  }
}
function slash(p) {
  return p.replace(windowsSlashRE, "/");
}
function normalizePath(id) {
  return import_path.default.posix.normalize(isWindows ? slash(id) : id);
}
function tryFsResolve(file) {
  const fileStat = tryStatSync(file);
  if (fileStat?.isFile()) {
    return normalizePath(file);
  }
  ;
  return;
}
function resolveAbsolutePath(rootDir, id, implicitExtension) {
  const file = import_path.default.join(rootDir, id);
  return tryFsResolve(file) || implicitExtension && implicitCivet(file) || tryFsResolve(id) || implicitExtension && implicitCivet(id);
}
function implicitCivet(file) {
  if (tryFsResolve(file)) {
    return;
  }
  const civet2 = file + ".civet";
  if (tryFsResolve(civet2)) {
    return civet2;
  }
  ;
  return;
}
var rawPlugin = (options = {}, meta) => {
  if (options.dts) options.emitDeclaration = options.dts;
  let compileOptions = {};
  let ts = options.ts;
  if (options.js) ts = "civet";
  if (!(ts != null)) {
    console.log('WARNING: You are using the default mode for `options.ts` which is `"civet"`. This mode does not support all TS features. If this is intentional, you should explicitly set `options.ts` to `"civet"`, or choose a different mode.');
    ts = "civet";
  }
  if (!["civet", "esbuild", "tsc", "preserve"].includes(ts)) {
    console.log(`WARNING: Invalid option ts: ${JSON.stringify(ts)}; switching to "civet"`);
    ts = "civet";
  }
  const transformTS = options.emitDeclaration || options.typecheck;
  const outExt = options.outputExtension ?? (ts === "preserve" ? ".tsx" : ".jsx");
  const implicitExtension = options.implicitExtension ?? true;
  let aliasResolver;
  let fsMap = /* @__PURE__ */ new Map();
  const sourceMaps = /* @__PURE__ */ new Map();
  let compilerOptions, compilerOptionsWithSourceMap;
  let rootDir = process.cwd();
  let outDir = import_path.default.join(rootDir, "dist");
  let esbuildOptions;
  let configErrors;
  let configFileNames;
  let skipWorker = false;
  let ref;
  if (transformTS || ts === "tsc") {
    ref = import("typescript").then(($1) => $1.default);
  } else {
    ref = void 0;
  }
  ;
  const tsPromise = ref;
  const getFormatHost = (sys) => {
    return {
      getCurrentDirectory: () => sys.getCurrentDirectory(),
      getNewLine: () => sys.newLine,
      getCanonicalFileName: sys.useCaseSensitiveFileNames ? (f) => f : (f) => f.toLowerCase()
    };
  };
  const cache = !(options.cache === false) ? /* @__PURE__ */ new Map() : void 0;
  const plugin = {
    name: "unplugin-civet",
    enforce: "pre",
    async buildStart() {
      let civetConfigPath = options.config;
      if (civetConfigPath === void 0) {
        civetConfigPath = await (0, import_config.findInDir)(process.cwd());
      }
      if (civetConfigPath) {
        compileOptions = await (0, import_config.loadConfig)(civetConfigPath);
      }
      compileOptions.parseOptions = {
        ...compileOptions.parseOptions,
        ...options.parseOptions
      };
      if (options.threads != null) {
        compileOptions.threads = options.threads;
      }
      if (transformTS || ts === "tsc") {
        let mogrify2 = function(key) {
          if (key in config && Array.isArray(config[key])) {
            return config[key] = config[key].map((item) => {
              if (!(typeof item === "string")) {
                return item;
              }
              return item.replace(/\.civet\b(?!\.)/g, ".civet.tsx");
            });
          }
          ;
          return;
        };
        var mogrify = mogrify2;
        const ts2 = await tsPromise;
        let config, error, tsConfigPath;
        if (options.tsConfig) {
          config = options.tsConfig;
          if (options.tsConfig.rootDir) {
            tsConfigPath = import_path.default.join(options.tsConfig.rootDir, "tsconfig.json");
          }
        } else {
          tsConfigPath = ts2.findConfigFile(process.cwd(), ts2.sys.fileExists);
          if (!tsConfigPath) {
            throw new Error("Could not find 'tsconfig.json'");
          }
          ;
          ({ config, error } = ts2.readConfigFile(
            tsConfigPath,
            ts2.sys.readFile
          ));
        }
        if (error) {
          console.error(ts2.formatDiagnostic(error, getFormatHost(ts2.sys)));
          throw error;
        }
        mogrify2("files");
        const system = { ...ts2.sys };
        const { readDirectory: systemReadDirectory } = system;
        system.readDirectory = (path2, extensions, excludes, includes, depth) => {
          extensions = [...extensions ?? [], ".civet"];
          return systemReadDirectory(path2, extensions, excludes, includes, depth).map(($2) => $2.endsWith(".civet") ? $2 + ".tsx" : $2);
        };
        const configContents = ts2.parseJsonConfigFileContent(
          config,
          system,
          tsConfigPath ? import_path.default.dirname(tsConfigPath) : process.cwd(),
          void 0,
          tsConfigPath
        );
        configErrors = configContents.errors;
        configFileNames = configContents.fileNames;
        compilerOptions = {
          ...configContents.options,
          target: ts2.ScriptTarget.ESNext,
          composite: false
        };
        compilerOptions.jsx ??= ts2.JsxEmit.Preserve;
        compilerOptionsWithSourceMap = {
          ...compilerOptions,
          sourceMap: true
        };
        fsMap = /* @__PURE__ */ new Map();
      }
    },
    async buildEnd(useConfigFileNames = false) {
      if (transformTS) {
        const ts2 = await tsPromise;
        const system = createFSBackedSystem(fsMap, process.cwd(), ts2);
        const {
          fileExists: systemFileExists,
          readFile: systemReadFile,
          readDirectory: systemReadDirectory
        } = system;
        system.fileExists = (filename) => {
          if (!filename.endsWith(".civet.tsx")) return systemFileExists(filename);
          if (fsMap.has(filename)) return true;
          return systemFileExists(filename.slice(0, -4));
        };
        system.readDirectory = (path2) => {
          return systemReadDirectory(path2).map(($3) => $3.endsWith(".civet") ? $3 + ".tsx" : $3);
        };
        const tsCompileOptions = {
          ...compileOptions,
          rewriteCivetImports: false,
          rewriteTsImports: true
        };
        system.readFile = (filename, encoding = "utf-8") => {
          if (import_path.default.basename(filename) === "package.json") {
            let recurse2 = function(node) {
              if (node != null && typeof node === "object") {
                for (const key in node) {
                  const value = node[key];
                  if (typeof value === "string") {
                    if (value.endsWith(".civet")) {
                      node[key] = value + ".tsx";
                      modified = true;
                    }
                  } else if (value) {
                    recurse2(value);
                  }
                }
              }
            };
            var recurse = recurse2;
            const json = systemReadFile(filename, encoding);
            if (!json) {
              return json;
            }
            const parsed = JSON.parse(json);
            let modified = false;
            recurse2(parsed.imports);
            return modified ? JSON.stringify(parsed) : json;
          }
          if (!filename.endsWith(".civet.tsx")) return systemReadFile(filename, encoding);
          if (fsMap.has(filename)) return fsMap.get(filename);
          const civetFilename = filename.slice(0, -4);
          const rawCivetSource = fs.readFileSync(civetFilename, { encoding });
          const { code: compiledTS, sourceMap } = import_civet.default.compile(rawCivetSource, {
            ...tsCompileOptions,
            filename,
            js: false,
            sourceMap: true,
            sync: true
            // TS readFile API seems to need to be synchronous
          });
          fsMap.set(filename, compiledTS);
          sourceMaps.set(filename, sourceMap);
          return compiledTS;
        };
        const host = createVirtualCompilerHost(
          system,
          compilerOptions,
          ts2
        );
        const program = ts2.createProgram({
          rootNames: useConfigFileNames ? configFileNames : [...fsMap.keys()],
          options: compilerOptions,
          host: host.compilerHost
        });
        const diagnostics = ts2.getPreEmitDiagnostics(program).map((diagnostic) => {
          const file = diagnostic.file;
          if (!file) return diagnostic;
          const sourceMap = sourceMaps.get(file.fileName);
          if (!sourceMap) return diagnostic;
          const sourcemapLines = sourceMap.lines ?? sourceMap.data.lines;
          const range = (0, import_ts_diagnostic.remapRange)(
            {
              start: diagnostic.start || 0,
              end: (diagnostic.start || 0) + (diagnostic.length || 1)
            },
            sourcemapLines
          );
          return {
            ...diagnostic,
            messageText: (0, import_ts_diagnostic.flattenDiagnosticMessageText)(diagnostic.messageText),
            length: diagnostic.length,
            start: range.start
          };
        });
        if (configErrors?.length) {
          diagnostics.unshift(...configErrors);
        }
        if (diagnostics.length > 0) {
          console.error(
            ts2.formatDiagnosticsWithColorAndContext(
              diagnostics,
              getFormatHost(ts2.sys)
            )
          );
          if (options.typecheck) {
            let failures = [];
            if (typeof options.typecheck === "string") {
              if (options.typecheck.includes("error")) failures.push(DiagnosticCategory.Error);
              if (options.typecheck.includes("warning")) failures.push(DiagnosticCategory.Warning);
              if (options.typecheck.includes("suggestion")) failures.push(DiagnosticCategory.Suggestion);
              if (options.typecheck.includes("message")) failures.push(DiagnosticCategory.Message);
              if (options.typecheck.includes("all")) {
                failures = { includes: () => true };
              }
            } else {
              failures.push(DiagnosticCategory.Error);
            }
            const count = diagnostics.filter((d) => failures.includes(d.category)).length;
            if (count) {
              const reason = count === diagnostics.length ? count : `${count} out of ${diagnostics.length}`;
              throw new Error(`Aborting build because of ${reason} TypeScript diagnostic${diagnostics.length > 1 ? "s" : ""} above`);
            }
          }
        }
        if (options.emitDeclaration) {
          if (meta.framework === "esbuild" && !esbuildOptions.outdir) {
            throw new Error("Civet unplugin's `emitDeclaration` requires esbuild's `outdir` option to be set;");
          }
          for (const file of fsMap.keys()) {
            const slashed = slash(file);
            if (!(file === slashed)) {
              fsMap.delete(slashed);
            }
          }
          for (const file of fsMap.keys()) {
            const sourceFile = program.getSourceFile(file);
            program.emit(
              sourceFile,
              (filePath, content) => {
                if (options.declarationExtension != null) {
                  if (filePath.endsWith(".d.ts")) {
                    filePath = filePath.slice(0, -5);
                  } else {
                    console.log(`WARNING: No .d.ts extension in ${filePath}`);
                  }
                  if (filePath.endsWith(civetSuffix)) {
                    filePath = filePath.slice(0, -civetSuffix.length);
                  } else {
                    console.log(`WARNING: No .civet extension in ${filePath}`);
                  }
                  filePath += options.declarationExtension;
                }
                let pathFromDistDir = import_path.default.relative(
                  compilerOptions.outDir ?? process.cwd(),
                  filePath
                );
                return this.emitFile({
                  source: content,
                  fileName: pathFromDistDir,
                  type: "asset"
                });
              },
              void 0,
              true,
              // emitDtsOnly
              void 0,
              // @ts-ignore @internal interface
              true
            );
          }
        }
      }
    },
    // forceDtsEmit
    resolveId(id, importer, options2) {
      if (aliasResolver != null) {
        id = aliasResolver(id);
      }
      if (/\0/.test(id)) return null;
      const { filename, postfix } = extractCivetFilename(id, outExt);
      let ref1;
      if (import_path.default.isAbsolute(filename)) {
        ref1 = resolveAbsolutePath(rootDir, filename, implicitExtension);
      } else {
        ref1 = import_path.default.resolve(import_path.default.dirname(importer ?? ""), filename);
      }
      ;
      let resolved = ref1;
      if (!resolved) return null;
      if (!resolved.endsWith(civetSuffix)) {
        if (!implicitExtension) return null;
        const implicitId = implicitCivet(resolved);
        if (!implicitId) return null;
        resolved = implicitId;
      }
      if (options2.scan && meta.framework === "vite") {
        resolved = `\0${resolved}`;
      }
      return resolved + (meta.framework === "esbuild" ? "" : outExt) + postfix;
    },
    loadInclude(id) {
      const { filename, postfix } = extractCivetFilename(id, outExt);
      if (skipWorker && workerRE.test(postfix)) {
        return false;
      }
      return filename.endsWith(civetSuffix);
    },
    async load(id) {
      let { filename } = extractCivetFilename(id, outExt);
      filename = import_path.default.resolve(rootDir, filename);
      this.addWatchFile(filename);
      let mtime, cached, resolve;
      if (cache != null) {
        try {
          mtime = (await fs.promises.stat(filename)).mtimeMs;
        } catch (e) {
        }
        if (mtime != null) {
          cached = cache.get(filename);
          if (cached && cached.mtime === mtime) {
            if (cached.promise) {
              await cached.promise;
            }
            let ref2;
            if ((ref2 = cached.result) != null) {
              const result = ref2;
              return result;
            }
          }
          const promise = new Promise((r) => {
            resolve = r;
          });
          cache.set(filename, cached = { mtime, promise });
        }
      }
      try {
        let checkErrors2 = function() {
          if (civetOptions.errors.length) {
            throw new import_civet.default.ParseErrors(civetOptions.errors);
          }
          ;
          return;
        };
        var checkErrors = checkErrors2;
        let compiled;
        let sourceMap;
        const civetOptions = {
          ...compileOptions,
          filename,
          errors: []
        };
        const rawCivetSource = (0, import_civet.decode)(await fs.promises.readFile(filename));
        const ast = await import_civet.default.compile(rawCivetSource, {
          ...civetOptions,
          ast: true
        });
        const civetSourceMap = new import_civet.SourceMap(
          rawCivetSource,
          normalizePath(import_path.default.relative(outDir, filename))
        );
        if (ts === "civet") {
          compiled = await import_civet.default.generate(ast, {
            ...civetOptions,
            js: true,
            sourceMap: civetSourceMap
          });
          sourceMap = civetSourceMap;
          checkErrors2();
        } else {
          const compiledTS = await import_civet.default.generate(ast, {
            ...civetOptions,
            js: false,
            sourceMap: civetSourceMap
          });
          checkErrors2();
          switch (ts) {
            case "esbuild": {
              const esbuildTransform = (await import("esbuild")).transform;
              const result = await esbuildTransform(compiledTS, {
                jsx: "preserve",
                loader: "tsx",
                sourcefile: filename,
                sourcemap: "external"
              });
              compiled = result.code;
              sourceMap = result.map;
              break;
            }
            case "tsc": {
              const tsTranspile = (await tsPromise).transpileModule;
              const result = tsTranspile(compiledTS, {
                compilerOptions: compilerOptionsWithSourceMap
              });
              compiled = result.outputText;
              sourceMap = result.sourceMapText;
              break;
            }
            case "preserve": {
              compiled = compiledTS;
              sourceMap = civetSourceMap;
              break;
            }
          }
        }
        if (transformTS) {
          for (let ref3 = import_civet.lib.gatherRecursive(ast, (($) => $.type === "ModuleSpecifier")), i = 0, len = ref3.length; i < len; i++) {
            const _spec = ref3[i];
            const spec = _spec;
            if (spec.module?.input) {
              spec.module.token = spec.module.input.replace(/\.([mc])?ts(['"])$/, ".$1js$2");
            }
          }
          const compiledTS = await import_civet.default.generate(ast, {
            ...civetOptions,
            js: false,
            sourceMap: civetSourceMap
          });
          checkErrors2();
          const tsx = filename + ".tsx";
          fsMap.set(tsx, compiledTS);
          sourceMaps.set(tsx, civetSourceMap);
          const slashed = slash(tsx);
          if (!(tsx === slashed)) {
            fsMap.set(slashed, compiledTS);
            sourceMaps.set(slashed, civetSourceMap);
          }
        }
        const jsonSourceMap = sourceMap && (typeof sourceMap === "string" ? JSON.parse(sourceMap) : sourceMap.json(normalizePath(import_path.default.relative(outDir, id))));
        let transformed = {
          code: compiled,
          map: jsonSourceMap
        };
        if (options.transformOutput) {
          transformed = await options.transformOutput(transformed.code, id);
        }
        if (cached != null) {
          cached.result = transformed;
          delete cached.promise;
        }
        return transformed;
      } finally {
        resolve?.();
      }
    },
    esbuild: {
      // Explicitly set the loader so esbuild handles the compiled output
      // correctly even though the virtual module id no longer ends in .jsx/.tsx.
      loader: ts === "preserve" ? "tsx" : "jsx",
      config(options2) {
        esbuildOptions = options2;
        outDir = options2.outdir || import_path.default.dirname(options2.outfile);
      }
    },
    vite: {
      config(config) {
        if (this?.meta?.viteVersion && 7 <= Number(this.meta.viteVersion.split(".")[0])) {
          skipWorker = true;
        }
        rootDir = import_path.default.resolve(process.cwd(), config.root ?? "");
        outDir = import_path.default.resolve(rootDir, config.build?.outDir ?? "dist");
        if (implicitExtension) {
          config.resolve ??= {};
          config.resolve.extensions ??= DEFAULT_EXTENSIONS;
          config.resolve.extensions.push(".civet");
        }
      },
      transformIndexHtml(html) {
        return html.replace(/<!--[^]*?-->|<[^<>]*>/g, (tag) => {
          return tag.replace(/<\s*script\b[^<>]*>/gi, (script) => {
            return script.replace(
              /([:_\p{ID_Start}][:\p{ID_Continue}]*)(\s*=\s*("[^"]*"|'[^']*'|[^\s"'=<>`]*))?/gu,
              (attr, name, value) => {
                return name.toLowerCase() === "src" && value ? attr.replace(
                  /(\.civet)(['"]?)$/,
                  (_, extension, endQuote) => {
                    return `${extension}${outExt}?transform${endQuote}`;
                  }
                ) : attr;
              }
            );
          });
        });
      },
      handleHotUpdate({ file, server, modules }) {
        if (!file.endsWith(".civet")) {
          return;
        }
        const resolvedId = slash(import_path.default.resolve(file) + outExt);
        let ref4;
        if (ref4 = server.moduleGraph.getModulesByFile(resolvedId)) {
          const fileModules = ref4;
          server.moduleGraph.onFileChange(resolvedId);
          return [...modules, ...fileModules];
        }
        return modules;
      }
    },
    rspack(compiler) {
      if (implicitExtension) {
        compiler.options ??= {};
        compiler.options.resolve ??= {};
        compiler.options.resolve.extensions ??= ["", ".js", ".json", ".wasm"];
        return compiler.options.resolve.extensions.unshift(".civet");
      }
      ;
      return;
    },
    webpack(compiler) {
      if (implicitExtension) {
        compiler.options.resolve ??= {};
        compiler.options.resolve.extensions ??= ["", ".js", ".json", ".wasm"];
        compiler.options.resolve.extensions.unshift(".civet");
      }
      return aliasResolver = (id) => {
        let ref5;
        for (const key in ref5 = compiler.options.resolve.alias) {
          const value = ref5[key];
          if (key.endsWith("$")) {
            if (id === key.slice(0, -1)) {
              return typeof value === "string" ? value : "\0";
            }
          } else {
            if (id === key || id.startsWith(key + "/")) {
              if (!(typeof value === "string")) {
                return "\0";
              }
              return value + id.slice(key.length);
            }
          }
        }
        return id;
      };
    }
  };
  return plugin;
};
var unplugin = (0, import_unplugin.createUnplugin)(rawPlugin);
var unplugin_default = unplugin;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  rawPlugin,
  slash
});
