// npm-check-updates@21.0.3/build/chunks/chunk-DvDhI3NP.js
import { createRequire as c } from "node:module";
var i = Object.create, _ = Object.defineProperty, O = Object.getOwnPropertyDescriptor, b = Object.getOwnPropertyNames, P = Object.getPrototypeOf, u = Object.prototype.hasOwnProperty, f = (e, r) => () => (e && (r = e(e = 0)), r), j = (e, r) => () => (r || e((r = { exports: {} }).exports, r), r.exports), x = (e, r) => {
  let t = {};
  for (var o in e)
    _(t, o, {
      get: e[o],
      enumerable: !0
    });
  return r || _(t, Symbol.toStringTag, { value: "Module" }), t;
}, n = (e, r, t, o) => {
  if (r && typeof r == "object" || typeof r == "function")
    for (var p = b(r), l = 0, s = p.length, a; l < s; l++)
      a = p[l], !u.call(e, a) && a !== t && _(e, a, {
        get: ((v) => r[v]).bind(null, a),
        enumerable: !(o = O(r, a)) || o.enumerable
      });
  return e;
}, w = (e, r, t) => (t = e != null ? i(P(e)) : {}, n(r || !e || !e.__esModule ? _(t, "default", {
  value: e,
  enumerable: !0
}) : t, e)), M = (e) => u.call(e, "module.exports") ? e["module.exports"] : n(_({}, "__esModule", { value: !0 }), e), S = /* @__PURE__ */ c(import.meta.url);
export {
  M as a,
  S as i,
  f as n,
  w as o,
  x as r,
  j as t
};
